Recording metrics for action result response statusCode in user-events service.

neerajmangal opened the pull request:
<!--- Provide a concise summary of your changes in the Title -->
This PR is the implementation of #4718, recording action's `statusCode` in Activation `event`, if it is set by the developer in action response. 


## Description
Currently, we can capture the `error` based on `Application`, `Developer` and `System`, with this change, the developer/operator can get visibility of actual status code sent in action response, i.e. 4XX, 5XX, etc by action. 
 
Action `statusCode` will be added in "Activation" `event` as `actionStatusCode` and consumed by user-event service. 

Note - For now `statusCode` present only if `statusCode` field is set by action developer.    

```json
"result": {
            "error": {
                "body": "Authentication credentials are invalid.",
                "statusCode": 401
            }
        }
```

With this change `Activation event`will look like this - 

```json
{
   "body": {
      "actionStatusCode": 404,
      "conductor": false,
      "duration": 46,
      "initTime": 34,
      "kind": "nodejs:10",
      "memory": 256,
      "name": "guest/helloWorld3",
      "size": 53,
      "statusCode": 1,
      "waitTime": 9
   },
   "eventType": "Activation",
   "namespace": "guest",
   "source": "invoker0",
   "subject": "guest",
   "timestamp": 1573811872958,
   "userId": "23bc46b1-71f6-4ed5-8c54-816aa4f8c502"
}
```

**Note** - `actionStatusCode` only present when - 

- Activation response has `error` field with `statusCode` set
- or Activation has `statusCode` set by developer.  
 
Sample Dashboard (added "Action Error Status Code" Panel) - 

<img width="1598" alt="Screenshot 2019-11-15 at 3 34 39 PM" src="https://user-images.githubusercontent.com/20740788/68934882-8fbd9080-07bd-11ea-9f8e-33105a89d791.png">

TODO
- [ ] Add panel to default dashboards. 

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [x]  I opened an issue to propose and discuss this change (#4718 )

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [x]  User events
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:
```diff
@@ -293,7 +295,9 @@ case class Activation(name: String,
                       causedBy: Option[String],
-                      size: Option[Int] = None)
+                      size: Option[Int] = None,
+                      actionStatusCode: Option[Int] = None
```
The inspection of the result of an action (instead of treating it opaquely) is a down side to this approach where otherwise the result is completely opaque. Naturally it would only work for actions that actually return an http response with a status code, which I don't think you can distinguish at the point of constructing a `Message`. Did you consider capturing the HTTP status code from the edge router instead for a /api/v1/web actions?

chetanmeh replied:
> The inspection of the result of an action (instead of treating it opaquely) is a down side to this approach where otherwise the result is completely opaque

The `statusCode` field is a documented structure for [web actions][1]. `statusCode` handling when under `error` field though is not documented properly. So current logic only looks for these documented structure to extract the info

> Did you consider capturing the HTTP status code from the edge router instead for a /api/v1/web actions

@rabbah If we capture the metrics at edge router then its tricky to distinguish a 5xx status from system error or web action actually responding with 5xx. Hence need to capture it at activation response level.

[1]: https://github.com/apache/openwhisk/blob/master/docs/webactions.md#handling-http-requests-with-actions

chetanmeh commented on tests/src/test/scala/org/apache/openwhisk/core/connector/test/EventMessageTests.scala:
```diff
@@ -90,5 +90,31 @@ class EventMessageTests extends FlatSpec with Matchers {
   }
 
+  it should "Transform a activation with error status code" in {
+    val resultWithError =
```
Also add one more variant with `statusCode` with `error`

chetanmeh commented on tests/src/test/scala/org/apache/openwhisk/core/connector/test/EventMessageTests.scala:
```diff
@@ -90,5 +90,31 @@ class EventMessageTests extends FlatSpec with Matchers {
+    val a =
+      fullActivation
+        .copy(response = ActivationResponse.applicationError(resultWithError.parseJson, Some(42)))
+    Activation.from(a) shouldBe Success(
```
Better to restrict the assertion to `statusCode` value itself instead of checking on whole assertion. Would reduce the noise

chetanmeh commented on common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:
```diff
@@ -340,7 +345,22 @@ object Activation extends DefaultJsonProtocol {
+    val statusCode = JsHelpers.getFieldPath(result.get.asJsObject, ERROR_FIELD, "statusCode")
+
+    statusCode match {
+      case Some(value) => Some(value.convertTo[Int])
```
Can be simplified  a bit here and also below
```scala
Some(JsNumber(code)) => Some(code.intValue())
```

rabbah replied:
the `statusCode` here is a use defined value - it could be of any type, including integer, string or other legal JSON values. Since this function is trying to interpret a use response it needs to be more defensive and also handle  `case _` when there are no matches.

also note that the controller handles string and int status codes.

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:
```diff
@@ -340,7 +345,22 @@ object Activation extends DefaultJsonProtocol {
+        JsHelpers.getFieldPath(result.get.asJsObject, "statusCode") match {
+          case Some(value) => Some(value.convertTo[Int])
+          case None        => None
+        }
```
```
statusCode
.map(_.intValue)
.orElse(_ => JsHelpers.getFieldPath(result.get.asJsObject, "statusCode").map(_.intValue))
```

i think - warning i didn't vet this in an ide/compiler but see my comment above, i don't think either of these are sufficient.

neerajmangal replied:
Sure, will try your suggestion.

selfxp commented on docs/metrics.md:
```diff
@@ -315,6 +315,7 @@ conductor - true for conductor backed actions
 memory - maximum memory allowed for action container
 causedBy - contains the "causedBy" annotation (can be "sequence" or nothing at the moment)
 size - size (in bytes) of the invocation response
+actionStatusCode - status code represents `statusCode` set in result response. (if not set, this field will not be present)
```
Should we rename the field into `httpCode` or `httpStatusCode` and point the [statusCode](https://github.com/apache/openwhisk/blob/master/docs/webactions.md#handling-http-requests-with-actions) to the documentation on how the user can set this up?

neerajmangal commented on core/monitoring/user-events/src/main/scala/org/apache/openwhisk/core/monitoring/metrics/KamonRecorder.scala:
```diff
@@ -34,6 +34,7 @@ trait KamonMetricNames extends MetricNames {
   val durationMetric = "openwhisk.action.duration"
   val responseSizeMetric = "openwhisk.action.responseSize"
   val statusMetric = "openwhisk.action.status"
+  val actionStatusCodeMetric = "openwhisk.action.errorCode"
```
Thanks for pointing this out @selfxp, I will change it accordingly. 
Initially, we thought to implement it only for web actions error code, but this is now applicable to all type actions if they have status code set in the response.

neerajmangal commented on core/monitoring/user-events/src/main/scala/org/apache/openwhisk/core/monitoring/metrics/PrometheusRecorder.scala:
```diff
@@ -46,6 +46,7 @@ trait PrometheusMetricNames extends MetricNames {
   val responseSizeMetric = "openwhisk_action_response_size_bytes"
   val statusMetric = "openwhisk_action_status"
   val memoryMetric = "openwhisk_action_memory"
+  val actionStatusCodeMetric = "openwhisk_action_errorcode"
```
Thanks! I missed it.

neerajmangal commented:
Thanks @rabbah, @chetanmeh and @selfxp for the feedback. Based on your feedback, I modified below - 

- More defensive checks when parsing the result response. 
  - if the result response contains an `error` field, check if it has `statusCode` defined by the user, as other fields are discarded if the `error` field is present in result response. 
     ```json
     "result": {
            "error": {
                "body": "Authentication credentials are invalid.",
                "statusCode": 401
            }
        }
     ````
  - else check if `statusCode` is set by the user without the `error` field in the result response.
     ```json
     "result": {
                "body": "Authentication credentials are invalid.",
                "statusCode": 401
        }
     ````
  - `statusCode` value is being checked for valid integer (401) and string value ("401").
  -  If there is any issue in getting the `intValue` (i.e. `statusCode` like "i401" or any string "xyz") return set `userDefinedStatusCode as `400` bad request. 
  - Using `value.convertTo[BigInt].intValue()` takes care of `JSNumber` and `JSString` type. I tried alternate approach like handling the status code for webaction https://github.com/apache/openwhisk/blob/master/core/controller/src/main/scala/org/apache/openwhisk/core/controller/WebActions.scala#L234, but for normal action use case it will miss a case where return `statusCode` is a valid integer (like "1") but not a http statusCode.
-  Changed `actionStatusCode` to `userDefinedStatusCode` as `statusCode` is not limited to http status codes. Please let me know if you have any better suggestions. 
-  Other tests and doc changes per feedback. 

Please let me know your feedback.

rabbah commented:
Changes lgtm. Sorry @neerajmangal for my delayed response. Can you resolve the conflict. Is there anything remaining on this PR?
